src/objects.c

```diff
@@ -1347,30 +1325,24 @@ static Obj FuncIS_POSOBJ(Obj self, Obj obj)
 */
 static Obj FuncSET_TYPE_POSOBJ(Obj self, Obj obj, Obj type)
 {
-#ifdef HPCGAP
     switch (TNUM_OBJ(obj)) {
-      case T_APOSOBJ:
-      case T_ALIST:
-      case T_FIXALIST:
-        SET_TYPE_OBJ( obj, type );
-        RetypeBag( obj, T_APOSOBJ );
-        CHANGED_BAG( obj );
-        break;
-      case T_POSOBJ:
-        SetTypePosObj( obj, type );
+#ifdef HPCGAP
+    case T_APOSOBJ:
+    case T_ALIST:
+    case T_FIXALIST:
+#endif
+    case T_POSOBJ:
         break;
-      default:
-        MEMBAR_WRITE();
-        SET_TYPE_POSOBJ(obj, type);
-        RetypeBag( obj, T_POSOBJ );
-        CHANGED_BAG( obj );
+    default:
+        if (!IS_PLIST(obj)) {
+            ErrorMayQuit("You can't make a positional object from a %s",
+                         (Int)TNAM_OBJ(obj), 0);
+        }
+        // TODO: we should also reject immutable plists, but that risks
+        // breaking existing code
```

Comment:
Hopefully this can be addressed in the future. I contemplated printing a warning, but it's too soon for that (see also PR #3851 and https://github.com/gap-packages/corelg/pull/8)